Sync state in current tab

guoyunhe opened the pull request:
Sync localStorage state in current browser tab by manually fire storage events.

Fix #28

franciscop reviewed (approved):
Looks good! Probably works fine since [it is already listening to this event](https://github.com/nas5w/use-local-storage/blob/master/index.ts#L82), right? Awesome to see this change merged in 🥳

franciscop reviewed (changes requested):
Actually a concern: won't this trigger two re-renders of the same hook? One update on the `setValue()`, and then dispatching this event will trigger another update on the hook listening for the event. Two solutions:
- We could probably at this point remove the whole local state and make it all based on the events.
- ~We could modify the setValue to trigger the event~ this would also result in two renders.

guoyunhe commented:
@franciscop indeed, need to add some sort of comparison.

guoyunhe commented:
@franciscop I have updated the PR. Introduced a ref called `rawValueRef` that store and compare raw localStorage values. So it won't update when the actual serialized content isn't change.

franciscop commented:
Oh it was much easier, you could have done this:

```js
if (e.newValue === serializer(storedValue)) return;
}, [key, syncData, storedValue]);
```

guoyunhe commented:
@franciscop I took performance into consideration. Using ref will reduce a lot of unnecessary serialization, which can be slow.

franciscop commented:
True! Code looks good 👍 , let's see what @nas5w says :)

franciscop commented on index.ts:
```diff
@@ -50,19 +54,40 @@ function useLocalStorage<T>(
+            url: window.location.href,
+            key,
+            newValue,
+            oldValue,
```
`oldValue` is not used anywhere?

guoyunhe replied:
@franciscop it is not used in useLocalStorage. But it exists in native StorageEvent. So it is better to have it, in case some other libraries need it.

nas5w commented:
Will review this shortly! Thank you guys!!

nas5w commented:
This looks awesome! I just want to take a step back and discuss this approach to make sure it won't have undesirable effects. 

As far as I know, there are three potential ways to accomplish state-syncing in the current tab for this hook:

**Manually fire a storage event as you have done in this PR**

The approach you have taken in this PR is clever! My only (very small) concern with it is it changes expected browser behavior: it's possible a user does not expect this browser event to be fired in the same tab. I'm not sure what the potential consequence of changing this behavior is. Do you think this could cause issues?

**Explicitly add context and require users to use a provider**

This would make it most obvious to your average React user what's going on. Generally, context is a familiar concept. However, this is a big paradigm shift and users may not be too happy having to add a provider to their app.

**Use an external, reactive store along with [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore)**

Avoids having to add context and also avoids changing browser behavior. However, I think this `useSyncExternalStore` hook is relatively new and may not support old React versions too well. Also, if not careful implementing an external store we could probably end up with memory leaks or other issues (I assume we'd want to use an external store library in this case).

---

Apologies for the brain-dump here, I just mostly want to make sure dispatching a Storage Event is the right direction and get your thoughts on the other options. Thanks!

franciscop commented:
"Use an external, reactive store along with [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore)" -> That's a narrow interpretation of "use an external store", which I would say should be the general way and using `useSyncExternalStore` is just an example implementation of it.

Basically all solutions are "use an store + an event system" (the source of truth + data sync between different parts), the nicest thing about @guoyunhe's solution is that it using `useState()` as the base store and the native browser events as the event system, no need to reinvent either.

I'll note the disadvantages of each approach:

- "Manually fire a storage event as you have done in this PR". What you noted basically, but I think it should be fine to fire the event even in the same window. I would have to think hard about the possibilities of an active conflict here with e.g. an extension.
- "Explicitly add context and require users to use a provider". We now need to add to the DX (Developer Experience) for the worse, and add a manual event system to sync the data up and down. IMHO since it's possible to solve the issue without worsening the DX, this is a no-go.
- "Use an external store with an event system". We'd have to implement both the external store system (or could use a fresh Redux's `createStore()`, but then we'd need to depend on Redux/similar) and the event system. Would work, but why go all the trouble?

There's a few other options, like I mentioned using an external store *library* like Redux, or also we could issue custom events like `selfstorage` instead of standard `storage` for the internal use-case. But I think overall chances of conflict with this library are so tiny that there's no need to worry. Heck, I'll go as far as to say I believe the spec is pretty bad in this case and it should trigger in the same window, maybe with a flag.

nas5w commented:
Okay, I agree. I’m going to do a bit of additional testing and merge in to main shortly/do a release. Thanks guys!
